pkg/hs/urbit-king/lib/Urbit/Vere/Ames.hs

```diff
@@ -194,36 +206,85 @@ ames env who isFake enqueueEv stderr = (initialEvents, runAmes)
     drv  <- mkRAcquire start stop
     pure (handleEffect drv mode)
 
-  start :: HasLogFunc e => RIO e AmesDrv
+  start :: RIO e AmesDrv
   start = do
+    mode <- rio (netMode isFake)
+    cache <- laneCache scryLane
+
     aTurfs   <- newTVarIO Nothing
     aDropped <- newTVarIO 0
+    aVersion <- newTVarIO Nothing
+    aVersTid <- trackVersionThread aVersion
     aUdpServ <- udpServ isFake who
-    aRecvTid <- queuePacketsThread aDropped aUdpServ
     aResolvr <- resolvServ aTurfs (usSend aUdpServ) stderr
+    aRecvTid <- queuePacketsThread
+      aDropped
+      aVersion
+      (byCache cache)
+      (send aUdpServ aResolvr mode)
+      aUdpServ
+
     pure (AmesDrv { .. })
 
   hearFailed _ = pure ()
 
-  queuePacketsThread :: HasLogFunc e => TVar Word -> UdpServ -> RIO e (Async ())
-  queuePacketsThread dropCtr UdpServ {..} = async $ forever $ do
-    outcome <- atomically $ do
-      (p, a, b) <- usRecv
-      enqueueEv (EvErr (hearEv p a b) hearFailed)
-    case outcome of
-      Intake -> pure ()
-      Ouster -> do
-        d <- atomically $ do
-          d <- readTVar dropCtr
-          writeTVar dropCtr (d + 1)
-          pure d
-        when (d `rem` packetsDroppedPerComplaint == 0) $
-          logWarn "ames: queue full; dropping inbound packets"
-
-  stop :: AmesDrv -> RIO e ()
+  trackVersionThread :: HasLogFunc e => TVar (Maybe Version) -> RIO e (Async ())
+  trackVersionThread versSlot = async $ forever do
+    env <- ask
+
+    scryVersion \v -> do
+      v0 <- readTVarIO versSlot
+      atomically $ writeTVar versSlot (Just v)
+      if (v0 == Just v)
+        then logInfo $ displayShow ("ames: proto version unchanged at", v)
+        else stderr ("ames: protocol version now " <> tshow v)
+
+    threadDelay (10 * 60 * 1_000_000)  -- 10m
+
+  queuePacketsThread :: HasLogFunc e
+                     => TVar Word
+                     -> TVar (Maybe Version)
+                     -> (Ship -> (Maybe [AmesDest] -> RIO e ()) -> RIO e ())
+                     -> (AmesDest -> ByteString -> RIO e ())
+                     -> UdpServ
+                     -> RIO e (Async ())
+  queuePacketsThread dropCtr vers lan forward UdpServ{..} = async $ forever $ do
+      -- port number, host address, bytestring
+    (p, a, b) <- atomically usRecv
+    ver <- readTVarIO vers
+
+    case decode b of
+      Right (pkt@Packet {..}) | ver == Nothing || ver == Just pktVersion -> do
+        logDebug $ displayShow ("ames: bon packet", pkt, showUD $ bytesAtom b)
+
+        if pktRcvr == who
+          then serf'sUp p a b
+          else lan pktRcvr $ \case
+            Just (dest:_) -> forward dest $ encode pkt
+              { pktOrigin = pktOrigin <|> Just (ipDest p a) }
+            _ -> logInfo $ displayShow ("ames: dropping unroutable", pkt)
+
+      Right pkt -> logInfo $ displayShow ("ames: dropping ill-versed", pkt, ver)
+
+      Left e -> logInfo $ displayShow ("ames: dropping malformed", e)
+
+    where
+      serf'sUp p a b =
```

Comment:
\>[serf's up](https://i.imgur.com/1rkt3Qo.png)

I know haskell syntax allows this, but is it considered fair game, style-wise, to put `'` in there? Wouldn't this be more legible as `enqueuePacket`, matching the `enqueueEv` it calls out to?

Reply:
It's not done commonly, but it's also not considered unstylish. Like if I saw a name with "dont" in it, I'd probably tell you to insert the apostrophe.

I get what you're saying with enqueue, but this name makes immediately clear where the thing is going, not merely that it is going into some queue. I also now reserve packet for `Packet`s and this thing takes a ByteString (you'll notice I also renamed `sendPacket` to `send` for this reason).